iceoryx_utils/source/posix_wrapper/timer.cpp

```diff
@@ -69,32 +69,73 @@ void Timer::OsTimer::callbackHelper(sigval data)
 
     auto& callbackHandle = OsTimer::s_callbackHandlePool[index];
 
-    // small optimazition to not lock the mutex if the callback handle is not valid anymore
+    // small optimization to not lock the mutex if the callback handle is not valid anymore
     if (descriptor != callbackHandle.m_descriptor.load(std::memory_order::memory_order_relaxed))
     {
         return;
     }
 
-    std::lock_guard<std::mutex> lock(callbackHandle.m_accessMutex);
-    if (!callbackHandle.m_inUse.load(std::memory_order::memory_order_relaxed))
+    uint64_t currentCycle = (callbackHandle.m_catchUpPolicy == CatchUpPolicy::SKIP_TO_NEXT_BEAT)
+                                ? 0u
+                                : callbackHandle.m_cycle.fetch_add(1u, std::memory_order_relaxed) + 1u;
+
+    // without the loop it is possible that a previous thread is interrupted right
+    // after the executeCallback call at the end and the succeeding thread skips the
+    // callback call since it is unable to acquire the mutex with try_lock.
+    // therefore we would skip a callback. to avoid this we have to make sure
+    // the callback is called by checking the callbackExecutionCycle in a while loop
+    if (callbackHandle.m_catchUpPolicy == CatchUpPolicy::IMMEDIATE)
     {
-        return;
-    }
+        while (currentCycle >= callbackHandle.m_callbackExecutionCycle.load(std::memory_order_relaxed))
```

Comment:
I already feared that something like this can happen and created an issue for that: https://github.com/eclipse/iceoryx/issues/170
For the moment I thought it would be robust enough since this issue should only arise under heavy load with a very small duration set but it seems that under Mac OS the problem already there...

Reply:
It is again refactored - hopefully now without any more mistakes